woocommerce/woocommerce-android
# Feature/9209 blaze tracking

Part of: #9209 


### Description

Adds tracking for Blaze webview flow by intercepting urls from the webview. This tracking required me to refactor how Blaze urls were loaded. I could not use the generic `WPComWebViewFragment` anymore and had to add a specific `BlazeWebViewFragment` alongside it's corresponding `BlazeWebViewViewModel` to be able to intercept URLs and track accordingly. This explains why a PR dedicated only for tracking is bigger than expected.   

The tracking is based on analytics plan from JP mobile, as documented in pe5sF9-1Ad-p2#metrics. This allows us to see the completion rate of the Blaze flow, and track how many campaigns are created on mobile.

The most nontrivial piece of the logic is how it calculates the current step of the Blaze flow given the URL, since JP mobile did some work on the web side to make it easier for tracking the steps like step-#. It's a bit hard to deduct what the step index means, though. In any case, I followed [the JP-Android implementation](https://github.com/wordpress-mobile/WordPress-Android/pull/18013/files#diff-38b882b0faf9a6f8b720f35e5b1482404e85ab1f545f7f8cd9a6ee59673ef7bf) for this with a reference link.

## Events added 

Event | Trigger | Properties
-- | -- | --
blaze_entry_point_displayed | When the Blaze entry point is shown to the user. | source *1
blaze_entry_point_tapped | When the Blaze entry point is tapped by the user. | source *1
blaze_flow_started | When the Blaze webview is first loaded. | source *1
blaze_flow_canceled | When the Blaze webview is dismissed without completing the flow. | source *1,current_step *2
blaze_flow_completed | When the Blaze webview is dismissed after completing the flow, when the step reaches step-5 according to the JPiOS Tracks event `jpios_blaze_flow_completed`. | source *1,current_step *2
blaze_flow_error | When the Blaze webview returns an error. | source *1,current_step *2



### Testing instructions

Open a site with blaze enabled any public atomic site.

- Go to the Menu tab --> the Blaze row should appear. When the Blaze row appears, an event should be logged `🔵 Tracked: blaze_entry_point_displayed, Properties: {"source":"menu","blog_id":205513046,"is_wpcom_store":false,"is_debug":true}`
- Tap on `Blaze` row --> an event should be logged `🔵 Tracked: blaze_entry_point_tapped, Properties: {"source":"menu","blog_id":205513046,"is_wpcom_store":false,"is_debug":true}`
- Navigate back to the Menu tab —> an event should be logged `🔵 Tracked: blaze_flow_canceled, Properties: {"source":"menu","step":"step-3","blog_id":205513046,"is_wpcom_store":false,"is_debug":true}`
- Tap on `Blaze` row again, and continue to complete creating a campaign (testing tips pe5sF9-1yI-p2#testing-tips) —> an event should be logged `🔵 Tracked: blaze_flow_completed, Properties: {"source":"menu","step":"step-5","blog_id":205513046,"is_wpcom_store":false,"is_debug":true}`

Repeat the same process starting Blaze from the Product detail screen -> overflow menu -> Promote with Blaze. This time there's no need to complete paying for the campaign as we already tested that in the previous steps

### Images/gif


https://github.com/woocommerce/woocommerce-android/assets/2663464/446a17eb-aa44-44df-80b7-406a395cdc40
